stix2/pattern_expressions.py

```diff
@@ -0,0 +1,223 @@
+
+    def __str__(self):
+        sub_exprs = []
+        for o in self.operands:
+            sub_exprs.append("%s" % o)
+        return (" " + self.operator + " ").join(sub_exprs)
+
+
+class AndObservableExpression(CompoundObservableExpression):
+    def __init__(self, operands):
+        super(AndObservableExpression, self).__init__("AND", operands)
+
+
+class OrObservableExpression(CompoundObservableExpression):
+    def __init__(self, operands):
+        super(OrObservableExpression, self).__init__("OR", operands)
+
+
+class FollowedByObservableExpression(CompoundObservableExpression):
+    def __init__(self, operands):
+        super(FollowedByObservableExpression, self).__init__("FOLLOWEDBY", operands)
+
+
+class ParentheticalExpression(PatternExpression):
+    def __init__(self, exp):
+        self.expression = exp
+        if hasattr(exp, "root_type"):
+            self.root_type = exp.root_type
+
+    def __str__(self):
+        return "(%s)" % self.expression
+
+
+class ExpressionQualifier(PatternExpression):
+    pass
+
+
+class RepeatQualifier(ExpressionQualifier):
+    def __init__(self, times_to_repeat):
+        if isinstance(times_to_repeat, IntegerConstant):
+            self.times_to_repeat = times_to_repeat
+        elif isinstance(times_to_repeat, int):
+            self.times_to_repeat = IntegerConstant(times_to_repeat)
+        else:
+            raise ValueError("%s is not a valid argument for a Within Qualifier" % times_to_repeat)
+
+    def __str__(self):
+        return "REPEATS %s TIMES" % self.times_to_repeat
+
+
+class WithinQualifier(ExpressionQualifier):
+    def __init__(self, number_of_seconds):
+        if isinstance(number_of_seconds, IntegerConstant):
+            self.number_of_seconds = number_of_seconds
+        elif isinstance(number_of_seconds, int):
+            self.number_of_seconds = IntegerConstant(number_of_seconds)
+        else:
+            raise ValueError("%s is not a valid argument for a Within Qualifier" % number_of_seconds)
+
+    def __str__(self):
+        return "WITHIN %s SECONDS" % self.number_of_seconds
+
+
+class StartStopQualifier(ExpressionQualifier):
+    def __init__(self, start_time, stop_time):
+        if isinstance(start_time, IntegerConstant):
+            self.start_time = start_time
+        elif isinstance(start_time, int):
+            self.start_time = IntegerConstant(start_time)
+        else:
+            raise ValueError("%s is not a valid argument for a Within Qualifier" % start_time)
+        if isinstance(stop_time, IntegerConstant):
+            self.stop_time = stop_time
+        elif isinstance(stop_time, int):
+            self.stop_time = IntegerConstant(stop_time)
+        else:
+            raise ValueError("%s is not a valid argument for a Within Qualifier" % stop_time)
+
+    def __str__(self):
+        return "START %s STOP %s" % (self.start_time, self.stop_time)
+
+
+class QualifiedObservationExpression(PatternExpression):
```

Comment:
Is this the most general class? Can you have qualifiers as part of a sub-expression?